packages/task/src/browser/task-definition-registry.ts

```diff
@@ -0,0 +1,65 @@
+/********************************************************************************
+ * Copyright (C) 2019 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable } from 'inversify';
+import {
+    TaskDefinition, TaskConfiguration, TaskDefinitionRegistry, TaskDefinitionContribution
+} from '../common';
+
+@injectable()
+export class TaskDefinitionRegistryImpl implements TaskDefinitionRegistry {
+
+    // task type - array of task definitions
+    private definitions: Map<string, TaskDefinition[]> = new Map();
+
+    getDefinitions(taskType: string): TaskDefinition[] {
+        return this.definitions.get(taskType) || [];
+    }
+
+    getDefinition(taskConfiguration: TaskConfiguration): TaskDefinition | undefined {
+        const definitions = this.getDefinitions(taskConfiguration.taskType || taskConfiguration.type);
+        let matchedDefinition: TaskDefinition | undefined;
+        let highest = -1;
+        for (const def of definitions) {
+            let score = 0;
+            if (!def.properties.required.every(requiredProp => taskConfiguration[requiredProp] !== undefined)) {
+                continue;
+            }
+            score += def.properties.required.length; // number of required properties
+            const requiredProps = new Set(def.properties.required);
+            // number of optional properties
+            score += def.properties.all.filter(p => !requiredProps.has(p) && taskConfiguration[p] !== undefined).length;
+            if (score > highest) {
+                highest = score;
+                matchedDefinition = def;
+            }
+        }
+        return matchedDefinition;
+    }
+
+    register(definitionContribution: TaskDefinitionContribution, pluginId: string): void {
```

Comment:
tas extension should not know anything about plugins

Reply:
i think this code should belong to plugin ext, task extension should not know anything about `TaskDefinitionContribution`, but register `TaskDefintion` instead

please review other apis to remove plugin contributions and replace them with internal task extension types

Reply:
i did half of the change and get stuck.

Conceptually it makes sense to move plugin contributions to `plugin-ext`, instead of having them in the `task` extension. 
However the plugin contributions don't always come from plugins. they could possibly be defined in the `tasks.json` by users to customise the tasks. and therefore, the `task` extension would need to have the knowledge of transforming `xxxContribution` to `xxx`. 
So, if i move the plugin contributions to the `plugin-ext`, i would need to duplicate some of them in the `task` extension. Otherwise there would be a circular dependency of `task` and `plugin-ext`.

Reply:
> task extension should not know anything about plugins

I agree. and like i explained above, some `XxxContribution` interfaces are actually not just for plugins. 
In my opinion, 
- `TaskDefinitionContribution` should renamed into `TaskDefinitionPluginContribution` be moved to `plugin-ext`
- Since `ProblemMatcherContribution`, `WatchingMatcherContribution`, & `ProblemPatternContribution` are for both `plug-ins` & `tasks.json`, they could stay in `task` extension

What do you think? @akosyakov

Reply:
sounds good, although i would get rid of `Contribution` suffix, and just call them `ProblemMatcher`, `WatchingWatcher` and `ProblemPattern`. They can be used as contributions in package.json but they seems to be used like definitions in tasks.json.

Reply:
in the most recent push, I updated the interfaces in a little different way.

I realise that the problemMatchers and problemPatterns from tasks.json and package.json differ slightly. 
- developers define problemMatchers and problemPatterns in package.json to **make contributions to the registries, so they always have a name**
- users of tasks.json **associate** problemMatchers and problemPatterns with defined tasks, and therefore the problemMatchers and problemPatterns in tasks.json **don't have to have a name**.

So, in my latest change, `PluginProblemMatcherContribution`, `PluginProblemPatternContribution`, and `PluginTaskDefinitionContribution` are defined in the `plugin-ext` for contributions from plugins. And `ProblemMatcherContribution`, `ProblemPatternContribution`, and `TaskDefinition` are defined in `tasks` for users of tasks.json.